segmentstore/server/src/main/java/io/pravega/segmentstore/server/logs/DataFrameReader.java

```diff
@@ -137,5 +170,87 @@ public DataFrameRecord<T> getNext() throws DataCorruptionException, DurableDataL
+     * @param logItem Internal information for this log item.
+     */
+    private void trackLastReadLogItem(long seqNo, DataFrameRecord.RecordInfo logItem) {
+        if (this.duplicateEntryTracker.containsKey(seqNo)) {
+            log.warn("{}: Found duplicate entry {}. Do not track it.", this.traceObjectId, seqNo);
+            return;
+        }
+
+        // We have a new valid log entry to track, just add it to the tracking map.
+        this.duplicateEntryTracker.put(seqNo, logItem);
+
+        // Remove exceeding elements from tracking map.
+        if (this.duplicateEntryTracker.getNumTrackedElements() > this.maxOverlapToCheckForDuplicates) {
+            this.duplicateEntryTracker.removeLast();
+        }
+    }
+
+    private boolean isHarmlessDuplicateEntry(long seqNo, DataFrameRecord.RecordInfo recordInfo) {
+        // Check if the duplicate comes from too far behind. In this case, we will report DCE anyway.
+        if (!duplicateEntryTracker.containsKey(seqNo)) {
+            log.warn("{}: Seems that we have found a duplicate {}, but it exceeds the max number of tracking elements {}.",
+                    this.traceObjectId, seqNo, this.maxOverlapToCheckForDuplicates);
+            return false;
+        }
+
+        // To determine if duplicate log entry is harmless, we use the available internal information of a DataFrame to
+        // check for equality: i) length of DataFrameRecord.EntryInfo list, and ii) compare that all DataFrameRecord.EntryInfo
+        // have the same length and offset within the DataFrame.
+        boolean harmlessDuplicate = duplicateEntryTracker.getEntriesFor(seqNo).size() == recordInfo.getEntries().size();
+        int scanUpTo = Math.min(duplicateEntryTracker.getEntriesFor(seqNo).size(), recordInfo.getEntries().size());
+        for (int i = 0; i < scanUpTo; i++) {
+            harmlessDuplicate &= compareDataFrameEntryInfos(duplicateEntryTracker.getEntriesFor(seqNo).get(i), recordInfo.getEntries().get(i));
+        }
+        return harmlessDuplicate;
+    }
+
+    /**
+     * Check if 2 repeated DataFrameRecord.EntryInfo items read from the log are equal comparing their offset and length
+     * within the DataFrame.
+     *
+     * @param originalEntryInfo Sequence number of the new log item to track.
+     * @param duplicateEntryInfo Internal information for this log item.
+     */
+    private boolean compareDataFrameEntryInfos(DataFrameRecord.EntryInfo originalEntryInfo, DataFrameRecord.EntryInfo duplicateEntryInfo) {
+        return originalEntryInfo.getFrameOffset() == duplicateEntryInfo.getFrameOffset()
+                && originalEntryInfo.getLength() == duplicateEntryInfo.getLength();
+    }
+
+    /**
+     * Helper class to keep track of recent read items from the log in order to check if it is safe to discard duplicate
+     * entries that could have been written to the log.
+     */
+    private static class DuplicateEntryTracker {
```

Comment:
If it's intended this should be annotated @NotThreadSafe.

Reply:
Sure.